ossl/src/asymcipher.rs

```diff
@@ -71,40 +71,90 @@ pub fn rsa_enc_params(
+pub enum EncOp {
+    Encrypt,
+    Decrypt,
+    Encapsulate,
+    Decapsulate,
+}
+
+/// Higher level wrapper for asymmetric cipher operations with OpenSSL
+///
+/// Supports Encryption, Decryption, Encapsulation, Decapsulation.
+/// Whether any of these operation will work also depends on the type
+/// of key provided.
+///
+/// An OsslError is returned if an operation is unsupported.
 #[derive(Debug)]
 pub struct OsslAsymcipher {
     /// The underlying OpenSSL EVP PKEY context.
     pkey_ctx: EvpPkeyCtx,
+    /// The requested operation type
+    op: EncOp,
 }
 
 impl OsslAsymcipher {
-    /// Initializes a new asymmetric encryption operation
-    pub fn message_encrypt_new(
+    /// Initializes a new asymmetric encryption or decryption operation
+    pub fn new(
         libctx: &OsslContext,
+        op: EncOp,
         key: &mut EvpPkey,
-        params: &OsslParam,
+        params: Option<&OsslParam>,
     ) -> Result<OsslAsymcipher, Error> {
         let mut ctx = OsslAsymcipher {
             pkey_ctx: key.new_ctx(libctx)?,
+            op: op,
         };
-        let ret = unsafe {
-            EVP_PKEY_encrypt_init_ex(ctx.pkey_ctx.as_mut_ptr(), params.as_ptr())
+        let params_ptr = match params {
+            Some(p) => p.as_ptr(),
+            None => std::ptr::null(),
+        };
+        let ret = match ctx.op {
+            EncOp::Encrypt => unsafe {
+                EVP_PKEY_encrypt_init_ex(ctx.pkey_ctx.as_mut_ptr(), params_ptr)
+            },
+            EncOp::Decrypt => unsafe {
+                EVP_PKEY_decrypt_init_ex(ctx.pkey_ctx.as_mut_ptr(), params_ptr)
+            },
+            EncOp::Encapsulate => unsafe {
+                EVP_PKEY_encapsulate_init(ctx.pkey_ctx.as_mut_ptr(), params_ptr)
+            },
+            EncOp::Decapsulate => unsafe {
+                EVP_PKEY_decapsulate_init(ctx.pkey_ctx.as_mut_ptr(), params_ptr)
+            },
         };
         if ret != 1 {
-            trace_ossl!("EVP_PKEY_encrypt_init()");
+            match ctx.op {
+                EncOp::Encrypt => {
+                    trace_ossl!("EVP_PKEY_encrypt_init()");
+                }
+                EncOp::Decrypt => {
+                    trace_ossl!("EVP_PKEY_decrypt_init()");
+                }
+                EncOp::Encapsulate => {
+                    trace_ossl!("EVP_PKEY_encapsulate_init()");
+                }
+                EncOp::Decapsulate => {
+                    trace_ossl!("EVP_PKEY_decapsulate_init()");
+                }
+            }
             return Err(Error::new(ErrorKind::OsslError));
         }
         Ok(ctx)
     }
 
     /// Executes an encryption operation if 'output' contains a buffer
     /// or probes and returns the expected output length if 'output' is None
-    pub fn message_encrypt(
+    pub fn encrypt(
         &mut self,
         plaintext: &[u8],
         output: Option<&mut [u8]>,
     ) -> Result<usize, Error> {
+        if self.op != EncOp::Encrypt {
```

Comment:
This error condition can be eliminated.  Similar to how we can use the type system to ensure that `OsslParam`s are always terminated, we can ensure that a well-typed program only invokes `encrypt` on `OsslAsymcipher`s that are configured for encryption.

Reply:
This could be done only by specializing in two separate structure, which would cause churn in kryoptic as it would have to hold multiple different handlers based on the operation.

I think this is a good compromise between ergonomicity and safety.